PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc

```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
 
   ONNXRuntime::~ONNXRuntime() {}
 
+  SessionOptions ONNXRuntime::defaultSessionOptions(GPUMode gpu_mode) {
+    SessionOptions sess_opts;
+    sess_opts.SetIntraOpNumThreads(1);
+    if (gpu_mode != no_gpu) {
+      // detect if there is GPU
+      int devices = 0;
+      auto status = cudaGetDeviceCount(&devices);
+      bool is_gpu_available = (status == cudaSuccess && devices > 0);
+      if (is_gpu_available) {
+        // https://www.onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html
+        OrtCUDAProviderOptions options;
```

Comment:
I see in https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L12-L14 and https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L23-L27
```spec
%if "%{cmsos}" != "slc7_aarch64"
Requires: cuda cudnn
%endif
...
%if "%{cmsos}" != "slc7_aarch64"
USE_CUDA=ON
%else
USE_CUDA=OFF
%endif
```
that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

Reply:
Given the condition `Requires: cuda` it would be better to make this package to explicitly depend on `cuda` external.

Reply:
> that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). 

The restriction here should be `aarch64` and not really `slc7` I think:)

> What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

Reply:
> > that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)).
> 
> The restriction here should be `aarch64` and not really `slc7` I think:)

Thanks, then we should fix that :) Let me tag @smuzaffar here too.

> > What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?
> 
> The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

Ok, then this code could be fine as a first attempt (I'd expect it would need to be changed at some point, but hard to guess now how), assuming `PhysicsTools/ONNXRuntime` would gain explicit dependence on `cuda` external.

Reply:
@hqucms @makortel , problem is with cudnn distribution for aarch64 for slc7. We use pre-build cudnn distributiuon ( https://github.com/cms-sw/cms-bot/blob/master/pr_testing/run-pr-comparisons#L124 ) . This was built for newer kernel ( and works on cs8) but failed for run/package for slc7. That is why the `cuda cudnn` is not included for `slc7_aarch64`.